Added lab for troubleshooting provisioning errors. (#140)
aws-samples/cfn101-workshop: workshop/content/basics/operations/troubleshooting-provisioning-errors/_index.md

---
title: 'Troubleshooting provisioning errors'
date: 2022-01-04T19:00:24Z
weight: 500
---

### Overview

> As you iterate on the development of your CloudFormation template, you test provisioning of resources you describe in your template by creating a CloudFormation [stack](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/stacks.html). If you specify incorrect property values for resource configurations in your template, by default your stack will roll back to the last-known stable state, and all stack resources will be rolled back. Let's take an example where you create a stack off of a template, in which you describe 10 resources. In this example, 9 resources you described are successfully created, and the creation of the tenth resource fails: by default, your stack will roll back, including the 9 resources that were successfully provisioned.

Reviewer: ```suggestion
As you iterate on the development of your CloudFormation template, you can test provisioning of resources described in your template by creating a CloudFormation [stack](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/stacks.html). If you specify incorrect property values for resource configurations in the template, by default the stack will roll back to the last-known stable state, and all stack resources will be rolled back. 

Let's take an example where you create a stack off of a template, in which you describe 10 resources. In this example, 9 resources are described as successfully created, and the creation of the tenth resource fails: by default, the stack will roll back, including the 9 resources that were successfully provisioned.
```

> To speed up development cycles, you can choose to [preserve](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/stack-failure-options.html) the state of resources that have been successfully provisioned as part of stack create and update operations, and of [change set](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-changesets.html) operations. When you choose to use this functionality, you pause the stack rollback to preserve the state of successfully provisioned resources, and you can focus on troubleshooting and fixing your configurations, so that you can resume your provisioning operations when you are ready.

Reviewer: ```suggestion
To speed up development cycles, you can choose to [preserve](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/stack-failure-options.html) the state of resources that have been successfully provisioned as part of stack create and update operations, and of [change set](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-changesets.html) operations. When you choose to use this functionality, the stack rollback is paused to preserve the state of successfully provisioned resources, then you can start troubleshooting and fixing your configurations, and resume provisioning operations when you are ready.
```

### Topics Covered
By the end of this lab, you will be able to:

> * Understand how to troubleshoot provisioning errors, whilst preserving the state resources you successfully deployed

Reviewer: ```suggestion
* Understand how to troubleshoot provisioning errors, whilst preserving the state resources successfully deployed
```

> * Navigate the [AWS resource and property types reference](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-template-resource-type-ref.html) to discover resource properties, and return values for a given resource, from which you can choose as you develop your template

Reviewer: ```suggestion
* Navigate to the [AWS resource and property types reference](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-template-resource-type-ref.html) to discover resource properties, and return values for a given resource
```

### Start Lab

> You will use an example CloudFormation template, that contains an incorrect resource configuration, to create a new stack. You will choose to *preserve successfully provisioned resources* as part of *stack failure options* to preserve the state of an example `ExampleDeadLetterQueue` [Amazon SQS](https://aws.amazon.com/sqs/) queue resource, that will be successfully provisioned as part of the stack creation operation.

Reviewer: ```suggestion
The example of CloudFormation template contains an incorrect resource configuration. To correct this, you will choose to *preserve successfully provisioned resources* as part of *stack failure options* to preserve the state of an example `ExampleDeadLetterQueue` [Amazon SQS](https://aws.amazon.com/sqs/) queue resource, to be successfully provisioned.
```

> The creation operation for your stack will fail because another SQS queue described in your template, `ExampleSourceQueue`, has a configuration error. You will troubleshoot and fix the error in the template, and then you will choose to resume the stack creation operation with the template you updated.

Reviewer: ```suggestion
The creation operation for the stack will fail because another SQS queue is described in the template, and has a configuration error. Once you troubleshoot and fix the error in the template, you can resume the stack creation operation with the updated template.
```

To get started, follow steps shown next:
1. Change directory to the `code/workspace/troubleshooting-provisioning-errors` directory.
2. Open the `example_sqs_queues.yaml` CloudFormation template in your favorite text editor.

> 3. Familiarize with the configuration for sample SQS queues in the template; your intents, in this example, are to:

Reviewer: ```suggestion
3. Familiarize yourself with the configuration of the sample SQS queues in the template; your intents, in this example, are to:
```

>     1. create a *source* queue and a [*dead-letter* queue](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-dead-letter-queues.html) (DLQ): you choose to have your `ExampleSourceQueue` target your `ExampleDeadLetterQueue` for messages that cannot be successfully processed.  In the template, you reference the [Amazon Resource Name](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) (ARN) for  `ExampleDeadLetterQueue` in the [redrive policy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html#aws-sqs-queue-redrive) for `ExampleSourceQueue`: your DLQ will then be created first, so that your *source* queue can then reference its ARN;

Reviewer: ```suggestion
    1. create a *source* queue and a [*dead-letter* queue](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-dead-letter-queues.html) (DLQ): have your `ExampleSourceQueue` target your `ExampleDeadLetterQueue` for messages that cannot be successfully processed.  In the template, reference the [Amazon Resource Name](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) (ARN) for  `ExampleDeadLetterQueue` in the [redrive policy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html#aws-sqs-queue-redrive) for `ExampleSourceQueue`: your DLQ will then be created first, so that your *source* queue can then reference its ARN;
```

>     2. describe both queues as [First-In-First-Out](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html#aws-sqs-queue-fifoqueue) (FIFO) queues in this example. When you describe a *source* queue and a *dead-letter* [queue](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html#aws-sqs-queue-redrive), both queues must be of the same type: either *standard* or *FIFO* (the latter is the example for this lab). You note that the example template contains different configuration values for the `FifoQueue` property of your 2 queues, for which you instead expect to have a value of `true` for both. Moreover, when you describe a FIFO queue, its name must contain a `.fifo` suffix: whilst `ExampleSourceQueue` contains the `.fifo` suffix in its `QueueName`, it is configured as a standard queue (i.e., with `FifoQueue` set to `false`), thus resulting in an error.

Reviewer: ```suggestion
    2. describe both queues as [First-In-First-Out](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html#aws-sqs-queue-fifoqueue) (FIFO) queues in this example. When you describe a *source* queue and a *dead-letter* [queue](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html#aws-sqs-queue-redrive), both queues must be of the same type: either *standard* or *FIFO* (the latter is the example for this lab). Note, that the example template contains different configuration for the two `FifoQueue` property values, instead of expected value of `true` for both. Moreover, when you describe a FIFO queue, its name must contain a `.fifo` suffix: whilst `ExampleSourceQueue` contains the `.fifo` suffix in its `QueueName`, it is configured as a standard queue (i.e., with `FifoQueue` set to `false`), thus resulting in an error.
```

> You will use the `example_sqs_queues.yaml` template, that contains the error mentioned earlier, to test the stack rollback pause functionality: you will then fix the error, and complete the stack creation next:

Reviewer: ```suggestion
Use the `example_sqs_queues.yaml` template, that contains the error mentioned earlier, to test the stack rollback pause functionality, then fix the error, and complete the stack creation:
```
